src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -67,22 +75,71 @@ public void execute(
-        for (int i = 0; i < querySearchResults.size(); i++) {
-            QuerySearchResult querySearchResult = querySearchResults.get(i);
-            if (!(querySearchResult.topDocs().topDocs instanceof CompoundTopDocs) || Objects.isNull(queryTopDocs.get(i))) {
-                continue;
-            }
-            CompoundTopDocs updatedTopDocs = queryTopDocs.get(i);
-            float maxScore = updatedTopDocs.totalHits.value > 0 ? updatedTopDocs.scoreDocs[0].score : 0.0f;
-            TopDocsAndMaxScore updatedTopDocsAndMaxScore = new TopDocsAndMaxScore(updatedTopDocs, maxScore);
+    private void updateOriginalQueryResults(
+        final List<QuerySearchResult> querySearchResults,
+        final FetchSearchResult fetchSearchResult,
+        final List<CompoundTopDocs> queryTopDocs,
+        final boolean isSingleShard
+    ) {
+        int queryTopDocsIndex = 0;
+        for (QuerySearchResult querySearchResult : querySearchResults) {
+            CompoundTopDocs updatedTopDocs = queryTopDocs.get(queryTopDocsIndex++);
+            float maxScore = updatedTopDocs.getTotalHits().value > 0 ? updatedTopDocs.getScoreDocs()[0].score : 0.0f;
+
+            // create final version of top docs with all updated values
+            TopDocs topDocs = new TopDocs(updatedTopDocs.getTotalHits(), updatedTopDocs.getScoreDocs());
+
+            TopDocsAndMaxScore updatedTopDocsAndMaxScore = new TopDocsAndMaxScore(topDocs, maxScore);
             querySearchResult.topDocs(updatedTopDocsAndMaxScore, null);
         }
+        // a workaround for a single shard case, fetch has happened, and we need to update both fetch and
+        // query results
+        if (isSingleShard && querySearchResults.size() == 1) {
+            updateFetchSearchResults(querySearchResults, fetchSearchResult);
+        }
+    }
+
+    private void updateFetchSearchResults(final List<QuerySearchResult> querySearchResults, final FetchSearchResult fetchSearchResult) {
+        if (Objects.isNull(fetchSearchResult)) {
+            return;
+        }
+        // fetch results have list of document content, that includes start/stop and
+        // delimiter elements. list is in original order from query searcher. We need to:
+        // 1. filter out start/stop and delimiter elements
+        // 2. filter out duplicates from different sub-queries
+        // 3. update original scores to normalized and combined values
+        // 4. order scores based on normalized and combined values
+        SearchHits searchHits = fetchSearchResult.hits();
+
+        // create map of docId to index of search hits
+        Map<Integer, SearchHit> docIdToSearchHit = new HashMap<>();
+        for (SearchHit searchHit : searchHits) {
+            docIdToSearchHit.put(searchHit.docId(), searchHit);
+        }
+        QuerySearchResult querySearchResult = querySearchResults.get(0);
```

Comment:
Why do we use only the first item in querySearchResults?

Reply:
for one shard we do have only one query result